File: content/bdd-with-cucumber/09/code/js/00-initial-commit/src/shouty.js
class Person {
  constructor(name, network, location) {
    this._name = name
    this._messages = []
    this._credits = 0
    this._network = network
    this._location = location

    this._network.subscribe(this)
  }

  get location() {
    return this._location
  }

  moveTo(location) {
    this._location = location
    return this
  }

  shout(message) {
    this._network.broadcast(message, this)
  }

  hear(message) {
    this._messages.push(message)
  }

  messagesHeard() {
    return this._messages
  }
}

class Network {
  constructor(range) {
    this._listeners = []
    this._range = range
  }

  subscribe(person) {
    this._listeners.push(person)
  }

  broadcast(message, shouter) {
    const shortEnough = message.length <= 180
    this._deductCredits(shortEnough, message, shouter)
    this._listeners.forEach((listener) => {
      const withinRange =
        Math.abs(listener.location - shouter.location) <= this._range
      if (withinRange && (shortEnough || shouter.credits >= 0)) {
        listener.hear(message)
      }
    })
  }

  _deductCredits(shortEnough, message, shouter) {
    if (!shortEnough) shouter.credits -= 2
    shouter.credits -= (message.match(/buy/gi) || []).length * 5
  }
}

module.exports = {
  Person,
  Network,
}

File: content/bdd-with-cucumber/09/code/ruby/10-extract-shout-steps/lib/shouty.rb
module Shouty
  class Person
    attr_reader :name, :messages_heard, :location
    attr_accessor :credits

    def initialize(name, network, location)
      @name           = name
      @messages_heard = []
      @network        = network
      @location       = location
      @credits        = 0

      @network.subscribe(self)
    end

    def shout(message)
      @network.broadcast(message, self)
    end

    def hear(message)
      @messages_heard << message
    end

    def move_to(new_location)
      @location = new_location
      self
    end
  end

  class Network
    attr_writer :range

    def initialize(range)
      @range     = range
      @listeners = []
    end

    def subscribe(person)
      @listeners << person
    end

    def broadcast(message, shouter)
      short_enough = message.size <= 180
      deduct_credits(short_enough, message, shouter)
      @listeners.each do |listener|
        within_range = (listener.location - shouter.location).abs <= @range

        if within_range && (short_enough || shouter.credits >= 0)
          listener.hear(message)
        end
      end
    end

    def deduct_credits(short_enough, message, shouter)
      shouter.credits -= 2 unless short_enough
      shouter.credits -= (message.scan(/buy/i) || []).size * 5
    end
  end
end
